src/blocksmith/plugins/FloatingTextFormatToolbarPlugin.js

```diff
@@ -0,0 +1,407 @@
+
+  const resetFormats = useCallback(() => {
+    setIsText(false);
+    setIsLink(false);
+    setIsBold(false);
+    setIsItalic(false);
+    setIsUnderline(false);
+    setIsStrikethrough(false);
+    setIsHighlight(false);
+    setBlockType('paragraph');
+  }, []);
+
+  const $updatePopup = useCallback(() => {
+    editor.getEditorState().read(() => {
+      const selection = $getSelection();
+
+      if (!$isRangeSelection(selection)) {
+        resetFormats();
+        return;
+      }
+
+      const node = getSelectedNode(selection);
+      const parent = node.getParent();
+
+      const isTextSelected = $isTextNode(node) || 
+        ($isTextNode(parent) && selection.getTextContent().length > 0);
+
+      if (!isTextSelected) {
+        resetFormats();
+        return;
+      }
+
+      setIsText(true);
+
+      setIsBold(selection.hasFormat('bold'));
+      setIsItalic(selection.hasFormat('italic'));
+      setIsUnderline(selection.hasFormat('underline'));
+      setIsStrikethrough(selection.hasFormat('strikethrough'));
+      setIsHighlight(selection.hasFormat('highlight'));
+
+      setIsLink($isLinkNode(parent) || $isLinkNode(node));
+
+      const anchorNode = selection.anchor.getNode();
+      let element = anchorNode.getKey() === 'root'
+        ? anchorNode
+        : $findMatchingParent(anchorNode, (e) => {
+          const parent = e.getParent();
+          return parent !== null && $isRootOrShadowRoot(parent);
+        });
+
+      if (element === null) {
+        element = anchorNode.getTopLevelElementOrThrow();
+      }
+
+      const elementKey = element.getKey();
+      const elementDOM = editor.getElementByKey(elementKey);
+      let newBlockType = element.getType();
+
+      if (elementDOM !== null && $isListNode(element)) {
+        const parentList = $getNearestNodeOfType(anchorNode, ListNode);
+        newBlockType = parentList ? parentList.getListType() : element.getListType();
+      }
+
+      setBlockType(newBlockType);
+    });
+  }, [editor, resetFormats]);
+
+  useEffect(() => {
+    document.addEventListener('selectionchange', $updatePopup);
+    return () => {
+      document.removeEventListener('selectionchange', $updatePopup);
+    };
+  }, [$updatePopup]);
+
+  useEffect(() => {
+    return mergeRegister(
+      editor.registerUpdateListener(() => {
+        $updatePopup();
+      }),
+      editor.registerCommand(
+        SELECTION_CHANGE_COMMAND,
+        () => {
+          $updatePopup();
+          return false;
+        },
+        COMMAND_PRIORITY_LOW
+      )
+    );
+  }, [editor, $updatePopup]);
+
+  return {
+    isText,
+    isLink,
+    isBold,
+    isItalic,
+    isUnderline,
+    isStrikethrough,
+    isHighlight,
+    blockType,
+  };
+}
+
+/**
+ * Based on Lexical's FloatingTextFormatToolbarPlugin
+ * @see https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/FloatingTextFormatToolbarPlugin/index.tsx
```

Comment:
@guizzettic link to the file as it was when you borrowed from it please. you can just follow the link in the comment and [press y](https://docs.github.com/en/repositories/working-with-files/using-files/getting-permanent-links-to-files#press-y-to-permalink-to-a-file-in-a-specific-commit)